File: A&DS/pcms8/task4.py
from sys import *
import threading

setrecursionlimit(10 ** 9)
threading.stack_size(2 ** 26)


def main():
    class Node:
        def __init__(self, group, availableways):
            self.availableways = availableways
            self.group = group

        def printnode(self):
            print(self.availableways, self.group)

    def dfs(node, congroup):
        node.group = congroup
        for i in range(len(node.availableways)):
            if infoaboutnodes[node.availableways[i]].group is None:
                dfs(infoaboutnodes[node.availableways[i]], congroup)

    fin = open("components.in", "r")
    fout = open("components.out", "w")

    matrixsize, lines = [int(x) for x in fin.readline().split()]
    infoaboutnodes = []

    for i in range(matrixsize):
        infoaboutnodes.append(Node(None, []))

    for i in range(lines):
        numofline, numofcol = [int(x) for x in fin.readline().split()]
        a = infoaboutnodes[numofline - 1].availableways
        b = infoaboutnodes[numofcol - 1].availableways
        if numofcol - 1 not in a:
            a.append(numofcol - 1)
        if numofline - 1 not in b:
            b.append(numofline - 1)

    cnt = 0
    for i in range(matrixsize):
        if infoaboutnodes[i].group is None:
            cnt += 1
            dfs(infoaboutnodes[i], cnt)

    print(cnt, file=fout)
    for node in infoaboutnodes:
        print(node.group, end=' ', file=fout)


thread = threading.Thread(target=main)
thread.start()

File: A&DS/pcms8/task4.cpp
#include <iostream>
#include <fstream>
#include <vector>
using namespace std;

void dfs(int node, int group, int infoaboutnodes[][2], vector<vector<int>>&availableways){
    int updnode;
    infoaboutnodes[node][0] = 1;
    infoaboutnodes[node][1] = group;
    for(int i = 0; i < availableways[node].size(); i++){
        updnode = availableways[node][i];
        if(infoaboutnodes[updnode][0] == 0){
            dfs(updnode, group, infoaboutnodes, availableways);
        }
    }
}

int main()
{
    ifstream fin("components.in");
    ofstream fout("components.out");

    int n, m, temp, cnt = 1;

    fin >> n >> m;

    int infoaboutnodes[n][2];
    vector<vector<int>> availableways(n, vector<int>());

    for(int i = 0; i < n; i++){
        infoaboutnodes[i][0] = 0;
        infoaboutnodes[i][1] = -1;
    }

    for(int i = 0; i < m; i++){
        int x, y;
        fin >> x >> y;
        availableways[x - 1].push_back(y - 1);
        availableways[y - 1].push_back(x - 1);
    }

    for(int i = 0; i < n; i++){
        if(infoaboutnodes[i][0] == 0){
            dfs(i, cnt, infoaboutnodes, availableways);
            cnt++;

        }
    }
    fout << cnt - 1 << endl;

    for(int i = 0; i < n; i++){
        fout << infoaboutnodes[i][1] << " ";
    }

    return 0;
}
